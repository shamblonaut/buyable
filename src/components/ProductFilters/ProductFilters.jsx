import { useEffect, useState } from "react";
import { useSearchParams } from "react-router-dom";
import PropTypes from "prop-types";

import { toTitleCase } from "@/utils/helpers";

import {
  CategoryControl,
  Container,
  ControlLabel,
  Controls,
  Dropdown,
  Input,
  SearchBar,
  SearchIcon,
  SortControl,
} from "./ProductFilters.styles";

const ProductFilters = ({ products, setFilteredProducts }) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const [searchQuery, setSearchQuery] = useState(
    searchParams.get("query") ?? "",
  );
  const [category, setCategory] = useState("all");
  const [sortBy, setSortBy] = useState("relevance");

  useEffect(() => {
    const categoryFilter = products.filter(
      (product) => category === "all" || product.category === category,
    );
    const searchFilter = categoryFilter.filter(
      (product) =>
        searchQuery === "" ||
        product.title.toLowerCase().includes(searchQuery.toLowerCase()) ||
        product.category.toLowerCase().includes(searchQuery.toLowerCase()),
    );

    let sortedProducts = searchFilter;
    switch (sortBy) {
      case "cheap":
        sortedProducts = searchFilter.sort(
          (productA, productB) => productA.price - productB.price,
        );
        break;
      case "expensive":
        sortedProducts = searchFilter.sort(
          (productA, productB) => productB.price - productA.price,
        );
        break;
      case "best":
        sortedProducts = searchFilter.sort(
          (productA, productB) => productB.rating.rate - productA.rating.rate,
        );
        break;
    }

    setFilteredProducts(sortedProducts);

    setSearchParams(searchQuery ? { query: searchQuery } : {});
  }, [
    products,
    searchQuery,
    category,
    sortBy,
    setSearchParams,
    setFilteredProducts,
  ]);

  const handleSearchInput = (event) => {
    setSearchQuery(event.target.value);
  };
  const handleCategoryChange = (event) => {
    setCategory(event.target.value);
  };
  const handleSortByChange = (event) => {
    setSortBy(event.target.value);
  };

  const getCategories = () => {
    const categories = [];
    for (const product of products) {
      if (!categories.includes(product.category)) {
        categories.push(product.category.toString());
      }
    }
    return categories;
  };

  return (
    <Container>
      <SearchBar>
        <SearchIcon />
        <Input
          type="text"
          aria-label="product-search-input"
          placeholder="Search Product"
          value={searchQuery}
          onChange={handleSearchInput}
        />
      </SearchBar>
      <Controls>
        <CategoryControl>
          <ControlLabel htmlFor="category-selector">Category: </ControlLabel>
          <Dropdown
            id="category-selector"
            aria-label="category-selector"
            value={category}
            onChange={handleCategoryChange}
          >
            <option value="all">All</option>
            {getCategories().map((category) => (
              <option value={category} key={category}>
                {toTitleCase(category)}
              </option>
            ))}
          </Dropdown>
        </CategoryControl>
        <SortControl>
          <ControlLabel htmlFor="sort-selector">Sort By: </ControlLabel>
          <Dropdown
            id="sort-selector"
            aria-label="sort-selector"
            value={sortBy}
            onChange={handleSortByChange}
          >
            <option value="relevance">Relevance</option>
            <option value="cheap">Price: Low to High</option>
            <option value="expensive">Price: High to Low</option>
            <option value="best">Best Rating</option>
          </Dropdown>
        </SortControl>
      </Controls>
    </Container>
  );
};

ProductFilters.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      image: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired,
      rating: PropTypes.exact({
        rate: PropTypes.number.isRequired,
        count: PropTypes.number.isRequired,
      }).isRequired,
    }).isRequired,
  ),
  setFilteredProducts: PropTypes.func.isRequired,
};

export default ProductFilters;
