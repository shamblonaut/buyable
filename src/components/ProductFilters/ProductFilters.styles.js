import { LucideSearch } from "lucide-react";
import styled from "styled-components";

export const Container = styled.div`
  width: 100%;

  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const Controls = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-width: 360px;
  width: 80%;
`;

const ControlField = styled.div`
  display: flex;
  gap: 8px;
  align-items: center;
  flex-wrap: wrap;
`;

export const SearchBar = styled.form`
  display: flex;
  align-items: center;
  margin: 16px;
  border: 1px solid grey;
  width: 80%;
  height: min-content;
  min-width: 360px;

  background-color: var(--color-light);
  border: 3px solid var(--color-dark);
  border-radius: 4px;
`;

export const SearchIcon = styled(LucideSearch)`
  margin: 8px;
`;

export const Input = styled.input`
  appearance: none;
  border: none;
  outline: none;
  background: none;
  color: inherit;

  flex: 1;
  padding: 8px;
  border-radius: inherit;
`;

export const CategoryControl = ControlField;

export const SortControl = styled(ControlField)`
  justify-content: flex-end;
`;

export const ControlLabel = styled.label`
  font-weight: 500;
`;

export const Dropdown = styled.select`
  color: var(--color-light);
  border: 3px solid var(--color-dark);
  border-radius: 4px;
  background-color: var(--color-accent);
  padding: 4px 8px;

  font-family: var(--font-highlight);
`;
