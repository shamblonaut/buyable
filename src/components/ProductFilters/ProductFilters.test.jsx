import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { BrowserRouter } from "react-router-dom";

import ProductFilters from "./ProductFilters";

const mockProducts = [
  {
    id: 1,
    price: 1023,
    title: "Proton Laptop 16GB RAM 512GB SSD",
    image: "http://example.com/image-1.jpg",
    category: "Electronics",
    rating: { rate: 3.5, count: 128 },
  },
  {
    id: 2,
    price: 511,
    title: "Tungstenite Graphics Card 6GB VRAM",
    image: "http://example.com/image-2.jpg",
    category: "Electronics",
    rating: { rate: 4.0, count: 64 },
  },
  {
    id: 3,
    price: 255,
    title: "Element X25 Smartphone",
    image: "http://example.com/image-3.jpg",
    category: "Electronics",
    rating: { rate: 2.3, count: 32 },
  },
  {
    id: 4,
    price: 127,
    title: "Electron Processor 2.4GHz",
    image: "http://example.com/image-4.jpg",
    category: "Electronics",
    rating: { rate: 4.8, count: 256 },
  },
  {
    id: 5,
    price: 63.49,
    title: "Nucleon Motherboard",
    image: "http://example.com/image-5.jpg",
    category: "Electronics",
    rating: { rate: 3.2, count: 16 },
  },
  {
    id: 6,
    price: 31.49,
    title: "Newton Sports Sneakers",
    image: "http://example.com/image-6.jpg",
    category: "Accessories",
    rating: { rate: 3.8, count: 96 },
  },
  {
    id: 7,
    price: 15.49,
    title: "Graviton Travel Bag",
    image: "http://example.com/image-7.jpg",
    category: "Accessories",
    rating: { rate: 2.7, count: 12 },
  },
  {
    id: 8,
    price: 7.99,
    title: "Torqus Chronometer Wristwatch",
    image: "http://example.com/image-8.jpg",
    category: "Accessories",
    rating: { rate: 2.1, count: 18 },
  },
  {
    id: 9,
    price: 3.99,
    title: "Pendulus Necklace",
    image: "http://example.com/image-9.jpg",
    category: "Accessories",
    rating: { rate: 1.8, count: 13 },
  },
  {
    id: 10,
    price: 1.99,
    title: "Centrifume Ring",
    image: "http://example.com/image-10.jpg",
    category: "Accessories",
    rating: { rate: 3.9, count: 84 },
  },
];

const mockSetFilteredProducts = vi.spyOn(
  {
    setFilteredProducts: vi.fn(),
  },
  "setFilteredProducts",
);

describe("ProductFilters component", () => {
  beforeEach(() => {
    vi.clearAllMocks();

    // Clear search queries
    history.replaceState(history.state, null, window.location.pathname);
  });

  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("shows all categories in category selector", () => {
    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    const selectorOptions = within(
      screen.getByRole("combobox", { name: "category-selector" }),
    ).getAllByRole("option");

    expect(selectorOptions[0].value).toBe("all");
    expect(selectorOptions[1].value).toBe("Electronics");
    expect(selectorOptions[2].value).toBe("Accessories");
  });

  it("modifies query url param on search", async () => {
    const user = userEvent.setup();

    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    await user.type(
      screen.getByRole("textbox", { name: "product-search-input" }),
      "product",
    );

    expect(window.location.search).toBe("?query=product");
  });

  it("filters products by matching search query with product titles", async () => {
    const user = userEvent.setup();

    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    await user.type(
      screen.getByRole("textbox", { name: "product-search-input" }),
      "gb",
    );

    expect(mockSetFilteredProducts).toHaveBeenCalledWith([
      mockProducts[0],
      mockProducts[1],
    ]);
  });

  it("filters products by matching search query with categories", async () => {
    const user = userEvent.setup();

    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    await user.type(
      screen.getByRole("textbox", { name: "product-search-input" }),
      "electronics",
    );

    expect(mockSetFilteredProducts).toHaveBeenCalledWith(
      mockProducts.slice(0, 5),
    );
  });

  it("filters by selected category", async () => {
    const user = userEvent.setup();

    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    await user.selectOptions(
      screen.getByRole("combobox", { name: "category-selector" }),
      "all",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith(mockProducts);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "category-selector" }),
      "Electronics",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith(
      mockProducts.slice(0, 5),
    );

    await user.selectOptions(
      screen.getByRole("combobox", { name: "category-selector" }),
      "Accessories",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith(
      mockProducts.slice(5, 10),
    );
  });

  it("filters by respective sorting methods", async () => {
    const user = userEvent.setup();

    const products = mockProducts.slice(0, 5);

    render(
      <BrowserRouter>
        <ProductFilters
          products={products}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "relevance",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith(products);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "cheap",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      products[4],
      products[3],
      products[2],
      products[1],
      products[0],
    ]);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "expensive",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      products[0],
      products[1],
      products[2],
      products[3],
      products[4],
    ]);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "best",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      products[3],
      products[1],
      products[0],
      products[4],
      products[2],
    ]);
  });

  it("adds together each filter to give a combined result", async () => {
    const user = userEvent.setup();

    render(
      <BrowserRouter>
        <ProductFilters
          products={mockProducts}
          setFilteredProducts={mockSetFilteredProducts}
        />
      </BrowserRouter>,
    );

    // Select electronics category
    await user.selectOptions(
      screen.getByRole("combobox", { name: "category-selector" }),
      "Electronics",
    );

    // Search for titles with "gb"
    await user.type(
      screen.getByRole("textbox", { name: "product-search-input" }),
      "gb",
    );

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "cheap",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      mockProducts[1],
      mockProducts[0],
    ]);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "expensive",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      mockProducts[0],
      mockProducts[1],
    ]);

    await user.selectOptions(
      screen.getByRole("combobox", { name: "sort-selector" }),
      "best",
    );

    expect(mockSetFilteredProducts).toHaveBeenLastCalledWith([
      mockProducts[1],
      mockProducts[0],
    ]);
  });
});
