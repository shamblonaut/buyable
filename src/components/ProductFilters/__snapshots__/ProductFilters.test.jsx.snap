// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductFilters component > renders correctly 1`] = `
<div>
  <div
    class="sc-bRKDuR bvLqoC"
  >
    <form
      class="sc-ggWZvA ebiOIG"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search sc-dTvVRJ gtrvVu"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
      <input
        aria-label="product-search-input"
        class="sc-jwTyAe blmavy"
        placeholder="Search Product"
        type="text"
        value=""
      />
    </form>
    <div
      class="sc-hvigdm dIyhsz"
    >
      <div
        class="sc-fhHczv dkZgVC"
      >
        <label
          class="sc-jJLAfE iFSvBN"
          for="category-selector"
        >
          Category: 
        </label>
        <select
          aria-label="category-selector"
          class="sc-hwkwBN bcSjSd"
          id="category-selector"
        >
          <option
            value="all"
          >
            All
          </option>
          <option
            value="Electronics"
          >
            Electronics
          </option>
          <option
            value="Accessories"
          >
            Accessories
          </option>
        </select>
      </div>
      <div
        class="sc-fhHczv sc-hjsuWn dkZgVC ddRyja"
      >
        <label
          class="sc-jJLAfE iFSvBN"
          for="sort-selector"
        >
          Sort By: 
        </label>
        <select
          aria-label="sort-selector"
          class="sc-hwkwBN bcSjSd"
          id="sort-selector"
        >
          <option
            value="relevance"
          >
            Relevance
          </option>
          <option
            value="cheap"
          >
            Price: Low to High
          </option>
          <option
            value="expensive"
          >
            Price: High to Low
          </option>
          <option
            value="best"
          >
            Best Rating
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
